### PR TITLE
Sentry issues

### PR DESCRIPTION
**Pull request by g4s8**
#1603 - added sentry interface, use fake implementation for tests and real sentry client for production

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1607](https://github.com/zerocracy/farm/issues/1607) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1607](https://github.com/zerocracy/farm/issues/1607) is assigned to @krzyk[/z](https://www.0crat.com/u/krzyk), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2290712569); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/sentry/SentryOf.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    private final Farm frm;
+
+    /**
+     * Ctor.
+     *
+     * @param farm Farm
+     */
+    public SentryOf(final Farm farm) {
+        this.frm = farm;
```
Comment (krzyk) [inline src/main/java/com/zerocracy/sentry/SentryOf.java]: @g4s8 there is no need for short field names (qulice doesn't enforce that now) - `farm` is enough

File: src/main/groovy/com/zerocracy/stk/pmo/profile/invite_friend.groovy [hunk 2]
```diff
@@ -64,7 +64,7 @@ def exec(Project pmo, XML xml) {
             { Project project -> new Roles(project).bootstrap().hasRole(author, 'QA') },
             farm.find('@id="C3NDPUA8L"')
           )
-        )
+        ),
```
Comment (krzyk) [inline src/main/groovy/com/zerocracy/stk/pmo/profile/invite_friend.groovy]: @g4s8 this doesn't look like related change

File: src/test/java/com/zerocracy/SafeSentryTest.java [hunk 3]
```diff
@@ -43,14 +42,10 @@ public void logsThrownException() throws Exception {
...
-            final SentryClient client = Mockito.mock(SentryClient.class);
-            Mockito.doThrow(new RuntimeException(message))
-                .when(client).sendException(Mockito.any(Exception.class));
             new SafeSentry(
-                client,
-                new PropsFarm(
-                    new Directives().xpath("/props/testing").remove()
-                )
+                (Sentry) error -> {
+                    throw new RuntimeException(message);
```
Comment (krzyk) [inline src/test/java/com/zerocracy/SafeSentryTest.java]: @g4s8 Maybe change this exception to e.g. `IllegalStateException`, this way you could get rid of PMD exclude and I think it should still work (it is a runtime exception after all).

**Comment by g4s8** [comment]
@krzyk fixed, please take a look

**Comment by krzyk** [comment]
@g4s8 thanks

**Comment by krzyk** [comment]
@rultor merge

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 11 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @krzyk[/z](https://www.0crat.com/u/krzyk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1607](https://github.com/zerocracy/farm/issues/1607) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by paulodamaso** [comment]
@g4s8 I'm still getting an error when trying to use sentry:
```
Error creating valid DSN from: 'https://7a4c0d4df4f24c73abfdc02646279d9d@sentry.io/1231395'.
io.sentry.dsn.InvalidDsnException: Invalid DSN, the following properties aren't set '[secret key]'
	at io.sentry.dsn.Dsn.validate(Dsn.java:211)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:65)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:40)
	at io.sentry.SentryClientFactory.resolveDsn(SentryClientFactory.java:73)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:42)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:21)
	at com.zerocracy.SafeSentry.<init>(SafeSentry.java:47)
	at com.zerocracy.farm.sync.Terminator.lambda$killer$1(Terminator.java:163)
	at org.cactoos.func.FuncOf.lambda$new$3(FuncOf.java:88)
	at org.cactoos.func.FuncOf.apply(FuncOf.java:104)
	at org.cactoos.func.UncheckedFunc.lambda$apply$0(UncheckedFunc.java:58)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:70)
	at org.cactoos.scalar.UncheckedScalar.value(UncheckedScalar.java:58)
	at org.cactoos.func.UncheckedFunc.apply(UncheckedFunc.java:59)
	at org.cactoos.func.RunnableOf.run(RunnableOf.java:89)
	at com.jcabi.log.VerboseRunnable.run(VerboseRunnable.java:190)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at com.jcabi.log.VerboseThreads$Wrap.run(VerboseThreads.java:221)
	at java.lang.Thread.run(Thread.java:748)
swallowed exception: io.sentry.dsn.InvalidDsnException: Invalid DSN, the following properties aren't set '[secret key]'
	at io.sentry.dsn.Dsn.validate(Dsn.java:211)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:65)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:40)
	at io.sentry.SentryClientFactory.resolveDsn(SentryClientFactory.java:73)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:42)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:21)
	at com.zerocracy.SafeSentry.<init>(SafeSentry.java:47)
	at com.zerocracy.farm.sync.Terminator.lambda$killer$1(Terminator.java:163)
	at org.cactoos.func.FuncOf.lambda$new$3(FuncOf.java:88)
	at org.cactoos.func.FuncOf.apply(FuncOf.java:104)
	at org.cactoos.func.UncheckedFunc.lambda$apply$0(UncheckedFunc.java:58)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:70)
	at org.cactoos.scalar.UncheckedScalar.value(UncheckedScalar.java:58)
	at org.cactoos.func.UncheckedFunc.apply(UncheckedFunc.java:59)
	at org.cactoos.func.RunnableOf.run(RunnableOf.java:89)
	at com.jcabi.log.VerboseRunnable.run(VerboseRunnable.java:190)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at com.jcabi.log.VerboseThreads$Wrap.run(VerboseThreads.java:221)
	at java.lang.Thread.run(Thread.java:748)
```
The curious is that these values are hardcoded and I can't find where.

**Comment by g4s8** [comment]
@paulodamaso it seems that you didn't merge master branch with this PR, because now `SafeSentry` doesn't call `SentryClientFactory.sentryClient` as in your stacktrace:
```
swallowed exception: io.sentry.dsn.InvalidDsnException: Invalid DSN, the following properties aren't set '[secret key]'
	at io.sentry.dsn.Dsn.validate(Dsn.java:211)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:65)
	at io.sentry.dsn.Dsn.<init>(Dsn.java:40)
	at io.sentry.SentryClientFactory.resolveDsn(SentryClientFactory.java:73)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:42)
	at io.sentry.SentryClientFactory.sentryClient(SentryClientFactory.java:21)
	at com.zerocracy.SafeSentry.<init>(SafeSentry.java:47)
```

**Comment by ypshenychka** [comment]
@krzyk According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only [one](https://github.com/zerocracy/farm/pull/1607#discussion_r210936171) major issue was found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by krzyk** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@krzyk thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**